File: T1 Q8 - Post-quantum/Kyber.py
import botan3 as botan
import os

def main():
    # Create a random number generator
    rng = botan.RandomNumberGenerator()

    # Generate a Kyber key pair
    kyber_key_size = 'Kyber-512-r3'  # Define Kyber parameters
    priv_key = botan.PrivateKey.create('Kyber', kyber_key_size, rng)
    pub_key = priv_key.get_public_key()

    # Encryptor setup with Kyber
    enc = botan.KemEncrypt(pub_key, 'HKDF(SHA-512)')
    salt = os.urandom(16)  # Generate a random salt
    shared_key_len = 32  # Length of shared key for AES
    kem_shared_key, encapped_key = enc.create_shared_key(rng, salt, shared_key_len)

    # Decryptor setup with Kyber
    dec = botan.KemDecrypt(priv_key, 'HKDF(SHA-512)')
    dec_shared_key = dec.decrypt_shared_key(salt, shared_key_len, encapped_key)

    # Verify the shared key
    if dec_shared_key != kem_shared_key:
        print("Shared keys differ")
        return 1

    # Use the shared key for AES encryption
    shared_key = kem_shared_key

    # Generate a 12-byte IV
    iv_size = 12
    iv = rng.get(iv_size)

    message = "This is a very private message"

    # Encrypt the message using AES-256/GCM
    cipher = botan.SymmetricCipher('AES-256/GCM')
    cipher.set_key(shared_key)
    cipher.start(iv)
    ciphertext = cipher.finish(message.encode('utf-8'))
    cipher_output = ciphertext.hex()
    print(f"_Encrypted (hex): {cipher_output}")

    # Decrypt the message using AES-256/GCM
    cipher = botan.SymmetricCipher('AES-256/GCM', False)
    cipher.set_key(shared_key)
    cipher.start(iv)
    plaintext = cipher.finish(bytes.fromhex(cipher_output))
    print(f"Decrypted output: {plaintext.decode('utf-8')}")

if __name__ == "__main__":
    main()

File: T1 Q8 - Post-quantum/Kyber.cpp
// Purpose: Kyber is a key encapsulation mechanism (KEM) used for securely exchanging symmetric keys.

#include <botan/kyber.h>
#include <botan/pubkey.h>
#include <botan/system_rng.h>
#include <botan/aead.h>
#include <array>
#include <iostream>

int main()
{
    const size_t shared_key_len = 32;
    const std::string kdf = "HKDF(SHA-512)";
    const std::string aead_algo = "AES-256/GCM";

    Botan::System_RNG rng;

    std::array<uint8_t, 16> salt;
    rng.randomize(salt);

    // Generate Kyber key pair
    Botan::Kyber_PrivateKey priv_key(rng, Botan::KyberMode::Kyber512_R3);
    auto pub_key = priv_key.public_key();

    // Encryptor setup
    Botan::PK_KEM_Encryptor enc(*pub_key, kdf);
    const auto kem_result = enc.encrypt(rng, shared_key_len, salt);

    // Decryptor setup
    Botan::PK_KEM_Decryptor dec(priv_key, rng, kdf);
    auto dec_shared_key = dec.decrypt(kem_result.encapsulated_shared_key(), shared_key_len, salt);

    if (dec_shared_key != kem_result.shared_key())
    {
        std::cerr << "Shared keys differ\n";
        return 1;
    }

    // Shared key to be used for AES encryption
    Botan::secure_vector<uint8_t> shared_key = kem_result.shared_key();

    // Example message to encrypt
    std::string plaintext = "This is a secret message.";

    // AEAD encryption
    std::unique_ptr<Botan::AEAD_Mode> encryptor = Botan::AEAD_Mode::create("AES-256/GCM", Botan::Cipher_Dir::Encryption);
    encryptor->set_key(shared_key);
    encryptor->start(salt);

    Botan::secure_vector<uint8_t> pt(plaintext.data(), plaintext.data() + plaintext.size());
    encryptor->finish(pt);

    std::cout << "_Encrypted (hex): ";
    for (auto c : pt)
    {
        std::cout << std::hex << static_cast<int>(c);
    }
    std::cout << std::endl;

    // AEAD decryption
    std::unique_ptr<Botan::AEAD_Mode> decryptor = Botan::AEAD_Mode::create(aead_algo, Botan::Cipher_Dir::Decryption);
    decryptor->set_key(shared_key);
    decryptor->start(salt);

    decryptor->finish(pt);

    std::string decrypted_text(reinterpret_cast<const char *>(pt.data()), pt.size());
    std::cout << "Decrypted output: " << decrypted_text << std::endl;

    return 0;
}
